platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,704 @@
+to that round.
+
+Each event must be validated using the roster associated with its birth-round. If Alice is far behind Bob, and she
+receives an event for a birth-round she doesn't have the roster for, then she cannot validate the event. So the Event
+Intake module will drop events with a birth-round too far in the future.
+
+If the event has been dropped, it is possible to fetch missing events from neighbors using Gossip or, if the node has
+fallen so far behind it can no longer fetch events from neighbors in gossip, then the node will reconnect.
+
+#### Self Events
+
+Events are not only given to the Event Intake system through gossip. Self events (those events created by the node
+itself) are also fed to Event Intake. These events **may** bypass some steps in the pipeline. For example, self-events
+do not need validation. Likewise, when replaying events from the pre-consensus event buffer, those checks are not needed
+(since they have already been proved valid and are in topological order).
+
+#### Neighbor Discipline
+
+During the validation process, invalid events are rejected, and this information is passed to the Sheriff module so the
+offending node may be disciplined. Note that the node to be disciplined will be the node that sent this bad event, not
+the creator. This information (which node sent the event) must be captured by Gossip and passed to Event Intake as part
+of the event metadata.
+
+#### Assigning Generations
+
+Several algorithms, including the hashgraph consensus algorithm and the tipset algorithm, require events to be part of
+a "generation". We use what is known as "local generation". Each node, when it starts, assigns "1" as the generation
+of the first event it encounters per event creator. The absolute numbers for the generations are not important, but
+their relative values are critical. The next generation will be the max generation of all parents, plus 1. So if Alice
+builds an event with parents `a4` and `b7`, the new event will be `a8`. Different nodes may assign different generations
+to the same events. The generation will be part of the event metadata, **not** part of `EventCore` (that is, it is not
+in the part of the event that is gossiped).
+
+Event Intake is responsible for assigning generations to events.
+
+#### Branch Detection
+
+The Event Intake module inspects events to determine whether any given event creator is "branching" the hashgraph. A
+"branch" happens when two or more different events from the same creator have the same "self-event" parent. Any node
+that branches (known affectionately as a "Dirty Rotten Brancher") will be reported to the Sheriff. Branching is a sign
+of either a dishonest node, or a seriously broken node. In either case, it may be subject to "shunning", and will be
+reported to the Execution layer for further observation and, if required, action (such as canceling rewards for
```

Comment:
Is DRB node first reported to the sheriff as a bad actor and then reported to the execution client?

Reply:
Correct. 

From the `Modules -> Gossip -> Neighbor Discipline` section: `If a neighbor misbehaves, the Gossip module will notify the Sheriff module that one of its neighbors is misbehaving.`

And from `Modules -> Sheriff Module`: `It also sends this information to the Execution layer`